### PR TITLE
fix android push notifications when app is closed

### PR DESCRIPTION
**Pull request by enahum**
#### Summary
On Android, when a push notification is received while the app is fully closed we were getting a crash with the following message

```
FATAL EXCEPTION: DefaultDispatcher-worker-1
Process: com.mattermost.rnbeta, PID: 12800
java.lang.IllegalAccessError: Interface com.nozbe.watermelondb.WMDatabase$TransactionFunction implemented by class com.mattermost.helpers.database_extension.GeneralKt$$ExternalSyntheticLambda2 is inaccessible (declaration of 'com.mattermost.helpers.database_extension.GeneralKt$$ExternalSyntheticLambda2' appears in /data/app/~~vkUbZhRNsPjUP2iOhNthEA==/com.mattermost.rnbeta-hrAYzoiHtA2oAM2UBe28qw==/base.apk!classes19.dex)
at com.mattermost.helpers.database_extension.GeneralKt.getDatabaseForServer(Unknown Source:0)
at com.mattermost.helpers.PushNotificationDataRunnable$Companion.start(PushNotificationDataHelper.kt:43)
at com.mattermost.helpers.PushNotificationDataHelper$fetchAndStoreDataForPushNotification$job$1.invokeSuspend(PushNotificationDataHelper.kt:23)
at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
    Suppressed: kotlinx.coroutines.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@4838933, Dispatchers.Default]
```

This is due to the fact that the WatermelonDB `interface TransactionFunction` was not explicitly set to public and for some reason not accessible while the app was not running, although it does not present the same problem when the app is in fact running either in the foreground or background.

Along with this fix, we migrated the CustomPushNotification class to kotlin for better interoperatibility ensuring we process only one notification at a time even when they arrive in quick succession.

#### Release Note
```release-note
Fixed a crash when receiving a push notification on Android when the app is not running
```

### CODE REVIEW

File: patches/@nozbe+watermelondb+0.27.1.patch [hunk 1]
```diff
@@ -414,17 +414,19 @@ index 2f170e0..bac0352 100644
...
          });
      }
  
+-    interface TransactionFunction {
 +    public void unsafeVacuum() {
 +        execute("vacuum");
 +    }
 +
-     interface TransactionFunction {
++    public interface TransactionFunction {
```
Comment (enahum) [inline patches/@nozbe+watermelondb+0.27.1.patch:426]: this is the actual fix

File: android/app/src/main/java/com/mattermost/helpers/database_extension/General.kt [hunk 2]
```diff
@@ -57,7 +57,7 @@ fun DatabaseHelper.getDatabaseForServer(context: Context?, serverUrl: String): W
             if (cursor.count == 1) {
                 cursor.moveToFirst()
                 val databasePath = String.format("file://%s", cursor.getString(0))
-                return WMDatabase.getInstance(databasePath, context!!)
+                return WMDatabase.buildDatabase(databasePath, context!!, SQLiteDatabase.CREATE_IF_NECESSARY)
```
Comment (enahum) [inline android/app/src/main/java/com/mattermost/helpers/database_extension/General.kt:60]: this is to ensure we create a new database connection as when they arrive in quick succession the code complains that we are trying to open an already closed database.

File: android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.kt
Comment (enahum) [inline android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.kt:1]: migrated to Kotlin

**Review by larkox — APPROVED** [review]
Great find!

**Comment by enahum** [comment]
I'd say this should even be cherry-picked for 2.17, thoughts?

cc: @larkox

**Comment by nobuto-m** [comment]
> I'd say this should even be cherry-picked for 2.17, thoughts?

Has this landed to 2.17 in the end? In the git history, it's slated for 2.18. However, the `v2.17.0` tag somehow contains the commit.

> $ git branch -rv --contains 38d3c7d1d1f7a270a368869eebc6f263c9c4dc7c
>   origin/build-release-528 b28cd2b1d Bump app build and version number (#8023) (#8024)
>   origin/release-2.18      b28cd2b1d Bump app build and version number (#8023) (#8024)

> $ git tag --contains 38d3c7d1d1f7a270a368869eebc6f263c9c4dc7c
> v2.17.0


I'm curious because I'm seeing instability of push notifications when the app is closed after upgrading from 2.16 to 2.17. If the patch is not included, I can test it out whenever the next beta gets available.

**Comment by devkartik123** [comment]
It is still not working. I just tried with the latest code.

**Pull request merged**